quickwit/quickwit-config/src/config_value.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::{any, fmt};
+
+use anyhow::{self, Context};
+use serde::{Deserialize, Deserializer};
+
+use crate::qw_env_vars::{QW_ENV_VARS, QW_NONE};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct ConfigValue<T, const E: usize> {
+    /// Value provided by the user in a config file.
+    provided: Option<T>,
+    /// Value provided by Quickwit as default.
+    default: Option<T>,
+}
+
+impl<T, const E: usize> ConfigValue<T, E>
+where
+    T: FromStr,
+    <T as FromStr>::Err: fmt::Debug,
+{
+    pub(crate) fn with_default(value: T) -> Self {
+        Self {
+            provided: None,
+            default: Some(value),
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn for_test(value: T) -> Self {
+        Self {
+            provided: Some(value),
+            default: None,
+        }
+    }
+
+    #[cfg(test)]
+    pub(crate) fn none() -> Self {
+        Self {
+            provided: None,
+            default: None,
+        }
+    }
+
+    #[cfg(any(test, feature = "testsuite"))]
+    pub(crate) fn unwrap(self) -> T {
+        self.provided.or(self.default).unwrap()
+    }
+
+    pub(crate) fn resolve_optional(
+        self,
+        env_vars: &HashMap<String, String>,
+    ) -> anyhow::Result<Option<T>> {
+        assert_eq!(QW_NONE, 0);
```

Comment:
Remove?

Reply:
I'd rather not because right below we do `if E > QW_NONE`. I really need `QW_NONE` to be `0`.

Reply:
I can make `QW_NONE` first on the list, add a comment, and a unit test.